crates/moonbuild/src/gen/gen_check.rs

```diff
@@ -257,24 +265,63 @@ pub fn gen_check(
     let _ = moonc_opt;
     let _ = moonbuild_opt;
     let mut dep_items = vec![];
-    for (_, pkg) in m.get_all_packages().iter() {
-        let item = pkg_to_check_item(&pkg.root_path, m.get_all_packages(), pkg, moonc_opt)?;
+
+    // --patch and --no-mi only apply over the package which specified by path
+    let mut spec_pkg_to_check = None;
+    let mut patch_path = None;
+    let mut no_mi = false;
+
+    // if pkg is specified, check that pkg and it's deps; if no pkg specified, check all pkgs
+    let pkgs_to_check = match moonbuild_opt.check_opt.as_ref() {
+        Some(CheckOpt {
+            package_path: Some(pkg_path),
+            patch_file: pp,
+            no_mi: nm,
+        }) => {
+            spec_pkg_to_check = m.get_package_by_path(&moonbuild_opt.source_dir.join(pkg_path));
+            patch_path = pp.clone();
+            no_mi = *nm;
+            &m.get_filtered_packages_and_its_deps(&moonbuild_opt.source_dir.join(pkg_path))
+        }
+        _ => m.get_all_packages(),
+    };
+
+    for (_, pkg) in pkgs_to_check {
+        let (no_mi_for_cur_pkg, patch_path_for_cur_pkg) = spec_pkg_to_check
+            .map(|p| {
+                if p.root_path == pkg.root_path {
+                    (no_mi, patch_path.clone())
+                } else {
+                    (false, None)
+                }
+            })
+            .unwrap_or((false, None));
+
+        let mut item = pkg_to_check_item(&pkg.root_path, m.get_all_packages(), pkg, moonc_opt)?;
+        item.patch_file = patch_path_for_cur_pkg.clone();
+        item.no_mi = no_mi_for_cur_pkg;
         dep_items.push(item);
+
         if !pkg.wbtest_files.is_empty() {
-            let item = pkg_with_wbtest_to_check_item(
+            let mut item = pkg_with_wbtest_to_check_item(
                 &pkg.root_path,
                 m.get_all_packages(),
                 pkg,
                 moonc_opt,
             )?;
+            item.patch_file = patch_path_for_cur_pkg.clone();
```

Comment:
Can we move this into `pkg_with_wbtest_to_check_item`?

Reply:
yes, but need to add extra args to `pkg_with_wbtest_to_check_item `

Reply:
Or try to move some dirty works into `run_check_internal` to keep `gen_check` simple and clean.